src/main/java/io/cdap/plugin/hubspot/source/streaming/HubspotReceiver.java

```diff
@@ -0,0 +1,90 @@
+ * the License.
+ */
+
+package io.cdap.plugin.hubspot.source.streaming;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.cdap.plugin.hubspot.common.HubspotHelper;
+import io.cdap.plugin.hubspot.common.HubspotPage;
+import io.cdap.plugin.hubspot.common.HubspotPagesIterator;
+import org.apache.spark.storage.StorageLevel;
+import org.apache.spark.streaming.receiver.Receiver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implementation of Spark receiver to receive Salesforce push topic events.
+ */
+public class HubspotReceiver extends Receiver<String> {
+  private static final Logger LOG = LoggerFactory.getLogger(HubspotReceiver.class);
+  private static final String RECEIVER_THREAD_NAME = "hubspot_api_listener";
+  private final HubspotStreamingSourceConfig config;
+
+  HubspotReceiver(HubspotStreamingSourceConfig config) throws IOException {
+    super(StorageLevel.MEMORY_AND_DISK_2());
+    this.config = config;
+  }
+
+  @Override
+  public void onStart() {
+    ThreadFactory namedThreadFactory = new ThreadFactoryBuilder()
+      .setNameFormat(RECEIVER_THREAD_NAME + "-%d")
+      .build();
+
+    Executors.newSingleThreadExecutor(namedThreadFactory).submit(this::receive);
+  }
+
+  @Override
+  public void onStop() {
+    // There is nothing we can do here as the thread calling receive()
+    // is designed to stop by itself if isStopped() returns false
+  }
+
+  private void receive() {
+    try {
+      HubspotPagesIterator hubspotPagesIterator = new HubspotPagesIterator(config);
+
+      while (!isStopped()) {
+        if (hubspotPagesIterator.hasNext()) {
+          store(hubspotPagesIterator.next().toString());
+        } else {
+          Integer minutesToSleep = config.getPullFrequency().getMinutesValue();
+          LOG.debug(String.format("Waiting for '%d' minutes to pull.", minutesToSleep));
+          Thread.sleep(TimeUnit.MINUTES.toMillis(minutesToSleep));
+
+          // reload current page
+          HubspotPage currentPage = new HubspotHelper().getHubspotPage(config,
+                                                                       hubspotPagesIterator.getCurrentPageOffset());
+          int iteratorPosition = hubspotPagesIterator.getIteratorPosition();
+
+          hubspotPagesIterator = new HubspotPagesIterator(config, currentPage,
+                                                          hubspotPagesIterator.getCurrentPageOffset());
+          hubspotPagesIterator.setIteratorPosition(iteratorPosition);
+        }
+      }
+    } catch (Exception e) {
+      String errorMessage = "Exception while receiving messages from hubspot";
+      // Since it's top level method of thread, we need to log the exception or it will be unseen
+      LOG.error(errorMessage, e);
+      throw new RuntimeException(errorMessage, e);
```

Comment:
The receiver will get terminated on error and we'll stop receiving messages. Can we retry (after some backoff) if the error is retryable?

Reply:
TODO

Reply:
The code is updated with the above TODO and linked to JIRA ticket . https://issues.cask.co/browse/PLUGIN-357